internal/action/tab.go

```diff
@@ -345,10 +347,16 @@ func (t *Tab) Resize() {
 }
 
 // CurPane returns the currently active pane
-func (t *Tab) CurPane() *BufPane {
-	p, ok := t.Panes[t.active].(*BufPane)
+func (t *Tab) CurPane() Pane {
```

Comment:
> Why not keep it returning `*BufPane`, and in the callers just check if the returned value is not nil?

This would be the easiest approach, but the function is called `CurPane()` and as such it should return the current pane, which might be a `BufPane`, `RawPane` (derived from `BufPane`), `InfoPane` (derived from `BufPane`) or a `TermPane`.
I just imagine the moment we really like to address the current active `TermPane` received from `CurPane()`, without changing the interface again.

> The lua-exported function returns `Pane`, yes, but according to the documentation it should return `*BufPane`. Isn't it better to fix it to be in accordance with the documentation?

But we could also correct the documentation, that the function returns a `Pane`. Anyway, isn't there a type conversion in Lua to address the `BufPane` based on the received value (I'm no Lua expert)?

Reply:
> Anyway, isn't there a type conversion in Lua to address the `BufPane` based on the received value (I'm no Lua expert)?

I'm afraid there isn't. I don't see https://pkg.go.dev/layeh.com/gopher-luar#New saying anything about Go's type assertion.

So, unless you know a way to do that, keeping the existing (implemented but not documented) polymorphic behavior lua's `CurState()` doesn't seem like a good idea. With the (documented but not implemented) non-polymorphic behavior, a plugin at least would be able to check if the current pane is a bufpane or not.

We may consider generalizing `CurPane()` to non-buf panes in the future, but I think that when it comes to non-buf panes, we have more important stuff to generalize anyway. At least the following:

- Lack of support for non-buf panes in chained actions. While this PR will fix the crashes in [#2307](https://github.com/zyedidia/micro/issues/2307) and [#2288](https://github.com/zyedidia/micro/issues/2288), it will still not make either of the keybindings in [#2307](https://github.com/zyedidia/micro/issues/2307) and [#2288](https://github.com/zyedidia/micro/issues/2288) actually work, since it just interrupts handling the action once it encounters to a non-buf pane.
- Lack of support for `onAction` and `preAction` callbacks for non-buf panes. See e.g. [2677#issuecomment-1409211651](https://github.com/zyedidia/micro/issues/2677#issuecomment-1409211651)

...In the meantime, when it comes to `CurPane()`:

- `InfoPane`: plugins can use `InfoBar().HasPrompt` to check if the infopane is currently active or not (like micro itself does).
- `TermPane`: yes, plugins probably cannot check if the current pane is a term pane (and which of them, if there are many of them), but there will never be a use case for that?
- `RawPane`: it is quite a special type of pane, plugins would hardly want to do anything with it.

Reply:
I more or less reverted the most of the stuff. I hope I didn't overlook or brake something in regards of plugin compatibility.
Currently it feels like not moving forward with `micro`, since so many functions are exported and can't simply changed or evolved. :/